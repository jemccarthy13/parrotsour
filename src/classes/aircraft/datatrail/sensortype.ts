export enum SensorType {
  RAW,
  ARROW,
}
