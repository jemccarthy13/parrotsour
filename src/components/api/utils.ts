export async function validAccessCode(accessCode: string) {
  const resp = await fetch(
    `${process.env.PUBLIC_URL}/database/codes/${accessCode}.php`,
    {
      method: "POST",
    }
  )

  return resp.status === 202
}
