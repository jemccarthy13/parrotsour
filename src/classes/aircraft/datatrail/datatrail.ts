import { Point } from "../../point"
import { IDMatrix } from "../id"

// TODO -- DATATRAIL -- register? so that Aircraft can generate one of each type of datatrail?
// TODO -- DATATRAIL -- Support a toggle for the different Sensor Types.....
export abstract class DataTrail {
  private startPos: Point

  constructor(startPos: Point) {
    this.startPos = startPos
  }

  public getStartPos(): Point {
    return this.startPos
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public move(_heading: number): void {
    //console.warn("Moving DataTrail in base class. Will have no effect.")
  }

  abstract getCenterOfMass(heading: number): Point

  abstract draw(heading: number, id: IDMatrix): void
}
