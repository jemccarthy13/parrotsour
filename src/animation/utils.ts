import { PaintBrush } from "../canvas/draw/paintbrush"
import { SensorType } from "../classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "../classes/groups/group"

/**
 * Internal function to see if a group is near the canvas boundary.
 *
 * Used to change group logic when approaching the edge of the canvas
 * @param group The group to check against boundaries
 * @param dataStyle The current DataTrail type
 */
export function isNearBounds(
  group: AircraftGroup,
  dataStyle: SensorType
): boolean {
  const buffer = 40
  const sX = group.getCenterOfMass(dataStyle).x
  const sY = group.getCenterOfMass(dataStyle).y
  const ctx = PaintBrush.getContext()

  return (
    sX < buffer ||
    sX > ctx.canvas.width - buffer ||
    sY < buffer ||
    sY > ctx.canvas.height - buffer
  )
}
