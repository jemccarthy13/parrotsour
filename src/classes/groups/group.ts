// Classes & Interfaces
import { SensorType } from "../aircraft/datatrail/sensortype"
import { Point } from "../point"
import { ACType, Aircraft } from "../aircraft/aircraft"
import { getMostRestrictiveID, IDMatrix } from "../aircraft/id"
import { IntentParams } from "../aircraft/intent"
import Tasking from "../taskings/tasking"
import { FORMAT } from "../supportedformats"

import { AltStack, getAltStack } from "../altstack"
// Functions
import {
  headingToRadians,
  PIXELS_TO_NM,
  randomNumber,
} from "../../utils/psmath"
import { drawGroupCap } from "./groupcap"
import { Braaseye } from "../braaseye"
import { Aspect, aspectFromCATA, toCardinal } from "../../utils/aspect"
import { BRAA } from "../braa"
import RangeBack from "./rangeback"

/**
 * The types of data that can be used to seed a group.
 *
 * Everything not provided is randomized or defaulted.
 */
export interface GroupParams {
  dataTrailType: SensorType
  sx?: number
  sy?: number
  nContacts?: number
  hdg?: number
  alts?: number[]
  desiredHdg?: number
  id?: IDMatrix
  type?: ACType
}

/**
 * An AircraftGroup is an array of 1+ Aircraft.
 *
 * It contains logic to move, give intent to, and draw Aircraft.
 *
 * TODO -- GROUP SPLIT/MERGE -- Currently maneuvers are handled as an all
 * or nothing. In the animator, consider additional logic for (man > 0), to update
 * "groups" var to split groups. Should be 'easy'
 */
export class AircraftGroup extends Array<Aircraft> {
  private braaseye!: Braaseye

  private startPos: Point = Point.DEFAULT
  private label = "GROUP"
  private useTrackDir = true
  private maneuvers = 0
  private anchor = false
  private useBull = false

  /**
   * Construct an AircraftGroup from the given parameters.Group will initialize
   * what it can, randomize what it can, and default the rest.
   *
   * @param p A Partial object of GroupParams.
   */
  constructor(p?: Partial<GroupParams>) {
    super()
    if (!p) p = {}
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // Seed params to pass into Aircraft
    p.sx = this.startPos.x
    p.sy = this.startPos.y
    p.hdg = p.hdg || 90

    this.startPos = new Point(p.sx, p.sy)

    // Create nContacts number of Aircraft and add to this collection
    const nContacts = p.nContacts || randomNumber(1, 4)
    for (let contact = 0; contact < nContacts; contact++) {
      if (p.alts && p.alts[contact])
        this.push(new Aircraft({ ...p, alt: p.alts[contact] }))
      else {
        this.push(new Aircraft(p))
      }

      // Compute start position offset for follow-on groups (calculated from
      // 90 deg perpendicular)
      const vectors = headingToRadians(p.hdg)
      p.sx += 2 * PIXELS_TO_NM * Math.cos(vectors.offset)
      p.sy += 2 * PIXELS_TO_NM * -Math.sin(vectors.offset)
    }

    // TODO -- MANEUVER -- % chance of 2 maneuvers (i.e. flank turn back hot)
    if (randomNumber(0, 100) < 20) {
      this.maneuvers = 1
    }
  }

  /*************************************************************************
   * Formatting
   *************************************************************************/
  setAnchor(b: boolean): void {
    this.anchor = b
    this.useBull = b
  }

  isAnchor(): boolean {
    return this.anchor
  }

  setUseBull(b: boolean): void {
    this.useBull = b
  }

  getUseBull(): boolean {
    return this.useBull
  }

  formatNumContacts(): string {
    let answer = ""
    if (this.getStrength() > 1) {
      if (this.getStrength() >= 3) {
        answer += "HEAVY "
      }
      answer += this.getStrength() + " CONTACTS"
    }
    return answer
  }

  /**
   * Return the string formatted answer for this group based on properties of the group
   */
  format(format: FORMAT, rangeBack?: RangeBack): string {
    // format label
    let answer = this.getLabel() + " "

    // format separation
    if (rangeBack) {
      answer += rangeBack.toString() + " "
    }

    // format bullseye if anchor priority
    const braaseye = this.getBraaseye()
    if (this.useBull || false) {
      answer += "BULLSEYE " + braaseye.bull.toString() + ", "
    }

    // format altitude stack
    const altStack = this.getAltStack(format)
    answer += altStack.stack + " "

    // format track direction
    const trackDir = this.getTrackDir()
    answer += " "
    answer += trackDir !== undefined ? trackDir : ""

    // apply ID
    answer += " HOSTILE "

    // apply fill-in for # contacts
    answer += this.formatNumContacts() + " "

    // apply fill-ins (HI/FAST/etc)
    answer += " " + altStack.fillIns
    return answer.replace(/ {2}/g, " ")
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this group's draw will start; will be approximately
   * the 'rear' end of the data trail.
   */
  getStartPos(): Point {
    return this[0].getStartPos()
  }

  /**
   * "Center of Mass" for an aircraft/group is one projected data plot ahead of the
   * data trail, averaged. For Aircraft, this is equiv. to one projected data plot ahead
   * of the Aircraft, on it's current heading.
   *
   * @returns The "center of mass" for this group, which is the average of each
   * Aircraft's center of mass.
   */
  getCenterOfMass(dataStyle: SensorType): Point {
    if (this.isCapping()) {
      return this.startPos
    }
    let x = 0
    let y = 0
    for (let idx = 0; idx < this.getStrength(); idx++) {
      const acPos = this[idx].getCenterOfMass(dataStyle)
      x += acPos.x
      y += acPos.y
    }
    return new Point(x / this.getStrength(), y / this.getStrength())
  }

  /**
   * @returns true iff one Aircraft in this group is capping
   */
  isCapping(): boolean {
    return (
      this.find((ac) => {
        return !ac.isCapping()
      }) === undefined
    )
  }

  /**
   * Set each aircraft in this group to be capping
   */
  setCapping(newVal: boolean): void {
    this.forEach((ac) => ac.setCapping(newVal))
  }

  setBraaseye(braaseye: Braaseye): void {
    this.braaseye = braaseye
  }

  getBraaseye(): Braaseye {
    return this.braaseye
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/

  /**
   * Get 'aspect' (HOT/FLANK/BEAM, etc) to another group
   *
   * Aspect is calculated by taking the angle difference between
   * other a/c heading, and the reciprocal bearing between this group
   * and other a/c.
   *
   * @param otherGrp other aircraft
   */
  getAspect(otherGrp: AircraftGroup, dataStyle: SensorType): Aspect {
    const recipBrg: BRAA = otherGrp
      .getCenterOfMass(dataStyle)
      .getBR(this.getCenterOfMass(dataStyle))

    let dist = (otherGrp.getHeading() - parseInt(recipBrg.bearing) + 360) % 360
    if (dist > 180) dist = 360 - dist
    const cata = dist

    return aspectFromCATA(cata)
  }

  /**
   * @returns Current heading of the Group. Currently assumes all Aicraft are
   * following the same heading, therefore only returns first Aircraft's heading.
   */
  getHeading(): number {
    return this[0].getHeading()
  }

  /**
   * @param newHdg New heading to set for all aircraft in this group
   */
  setHeading(newHdg: number): void {
    this.forEach((ac) => ac.setHeading(newHdg))
  }

  /**
   * @returns Cardinal track direction or undefined. If there is a picture
   * track direction specified (i.e. all groups track West) this function
   * returns undefined. Otherwise, cardinal direction from heading.
   */
  getTrackDir(): string | undefined {
    let trackDir = undefined
    if (this.isCapping()) {
      trackDir = "CAP"
    } else if (this.useTrackDir) {
      trackDir = "TRACK " + toCardinal(this.getHeading())
    }
    return trackDir
  }

  /**
   * Set the useTrackDir flag, whether or not to include track direction
   * in group formatting
   */
  setUseTrackDir(newVal: boolean): void {
    this.useTrackDir = newVal
  }

  /**
   * Update maneuver count
   * @param newNumManeuvers New number of maneuvers remaining
   */
  setManeuvers(newNumManeuvers: number): void {
    this.maneuvers = newNumManeuvers
  }

  /**
   * @returns true iff group has at least one maneuver left.
   */
  doesManeuvers(): boolean {
    return this.maneuvers > 0
  }

  /*************************************************************************
   * Altitude(s)
   *************************************************************************/
  /**
   * @returns Altitude of the highest Aircraft in this group
   */
  getAltitude(): number {
    return Math.max(...this.getAltitudes())
  }

  /**
   * @param format The desired formatting
   * @returns AltStack containing formatted STACKS/Fillins
   */
  getAltStack(format: FORMAT): AltStack {
    return getAltStack(this.getAltitudes(), format)
  }

  /**
   * @returns number[] containing the altitude of every Aircraft in this Group
   */
  getAltitudes(): number[] {
    const alts = []
    for (let idx = 0; idx < this.length; idx++) {
      alts.push(this[idx].getAltitude())
    }
    return alts
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current group label
   */
  getLabel(): string {
    return this.label
  }

  /**
   * @param newLbl New group label
   */
  setLabel(newLbl: string): void {
    this.label = newLbl
  }

  /**
   * Update group ID
   * @param newID
   */
  setIDMatrix(newID: IDMatrix): void {
    this.forEach((ac) => {
      ac.setIDMatrix(newID)
    })
  }

  /**
   * @returns The most restrictive ID of child Aircraft
   */
  getIDMatrix(): IDMatrix {
    const ids: IDMatrix[] = []
    this.forEach((ac) => {
      ids.push(ac.getIDMatrix())
    })
    return getMostRestrictiveID(ids)
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Number of contacts in this group
   */
  getStrength(): number {
    return this.length
  }

  /**
   * @returns The Type of Aircraft in this group. Currently only supports one type
   * for the entire group (first aircraft)
   */
  getType(): ACType {
    return this[0].getType()
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw each aircraft. See Aircraft.draw(...)
   * @param context Current drawing context
   * @param dataStyle The type of DataTrail to use
   */
  draw(dataStyle: SensorType): void {
    if (this.isCapping()) {
      drawGroupCap(this)
    } else {
      this.forEach((ac) => ac.draw(dataStyle))
    }
  }

  /**
   * Move each Aircraft. See Aircraft.move()
   */
  move(): void {
    this.forEach((ac) => ac.move())
  }

  /**
   * Update each Aircraft's altitude based on intent. See Aircraft.doNextAltChange()
   */
  updateAltitude(): void {
    this.forEach((ac) => ac.doNextAltChange())
  }

  /*************************************************************************
   * Routing
   *************************************************************************/
  /**
   * Update intent for each Aircraft in the group. See Aircraft.updateIntent()
   * @param newIntent Object representing new intent for each Aircraft.
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.forEach((ac) => ac.updateIntent(newIntent))
  }

  /**
   * @returns true iff there is an Aircraft with intended destination
   */
  hasRouting(): boolean {
    return this.find((ac) => !ac.atFinalDestination()) !== undefined
  }

  /**
   * @returns The next Point on this group's intended route. Assumes
   * all Aircraft in the Group are following the same route, therefore returns
   * only the first Aircraft's next routing point.
   */
  getNextRoutingPoint(): Point | undefined {
    return this[0].getNextRoutingPoint()
  }

  /**
   * @param pt Point to add to all group Aircraft(s) routes
   */
  addRoutingPoint(pt: Point): void {
    this.forEach((ac) => ac.addRoutingPoint(pt))
  }

  /**
   * Perform the next routing action for each aircraft. See Aircraft.doNextRouting()
   */
  doNextRouting(): void {
    this.forEach((ac) => ac.doNextRouting())
  }

  /**
   * Clears all currently assigned transit locations
   */
  clearRouting(): void {
    this.forEach((ac) => ac.clearRouting())
  }

  /*************************************************************************
   * Tasking
   *************************************************************************/
  /**
   * @param task Aircraft's new Tasking.
   */
  setTasking(task: Tasking): void {
    this.forEach((ac: Aircraft) => {
      ac.setTasking(task)
    })
  }

  /**
   * @returns true if there is at least one Aircraft in this group that has a Tasking.
   */
  isOnTask(): boolean {
    return this.find((ac) => ac.isTasked()) !== undefined
  }
}
