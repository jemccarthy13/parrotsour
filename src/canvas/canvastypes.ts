import { Point } from "classes/point"
import { FORMAT } from "../classes/supportedformats"
import { SensorType } from "classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "classes/groups/group"

export enum BlueInThe {
  NORTH,
  SOUTH,
  EAST,
  WEST,
}

export class FightAxis {
  public static isNS(orientation: BlueInThe): boolean {
    return orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
  }
}

export class CanvasOrient {
  height = 0
  width = 0
  orient = BlueInThe.EAST
}

export interface CanvasProps {
  orientation: CanvasOrient
  picType: string
  braaFirst: boolean
  dataStyle: SensorType
  showMeasurements: boolean
  isHardMode: boolean
  newPic: boolean
  animate: boolean
  resetCallback?: () => void
  animateCallback: () => void
}

export interface PictureCanvasState {
  bullseye: Point
  blueAir: AircraftGroup
  answer: PictureAnswer
  reDraw: PictureReDrawFunction
  animateCanvas?: ImageData
}

export interface CanvasDrawFunction {
  (context: CanvasRenderingContext2D): Promise<void>
}

export interface DrawCanvasProps extends CanvasProps {
  draw: CanvasDrawFunction
  bullseye: Point
  answer: PictureAnswer
}

export interface PictureCanvasProps extends CanvasProps {
  format: FORMAT
  setAnswer: { (answer: PictureAnswer): void }
  sliderSpeed: number
  desiredNumContacts: number
}

export type PictureAnswer = {
  pic: string
  groups: AircraftGroup[]
}

export interface PictureDrawFunction {
  (
    context: CanvasRenderingContext2D,
    props: PictureCanvasProps,
    state: PictureCanvasState,
    hasCaps: boolean,
    desiredNumContacts: number,
    start?: Point
  ): PictureAnswer
}

export interface PictureReDrawFunction {
  (forced?: boolean, start?: Point, hasCaps?: boolean): PictureAnswer
}
