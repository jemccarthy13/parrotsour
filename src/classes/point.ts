import { BlueInThe } from "../canvas/canvastypes"
import { BRAA } from "./braa"
import { PIXELS_TO_NM, toDegrees } from "../utils/psmath"

/**
 * An x,y Cartesian point that can calculate bearing and
 * range to another point.
 */
export class Point {
  x: number
  y: number

  static DEFAULT = new Point(0, 0)

  /**
   * Construct a new x,y point
   */
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  /**
   * Get a bearing and range between this point and another point
   *
   * @param ptTo the "other" or "to" point
   * @returns the bearing and range between this (ptFrom) and ptTo
   */
  getBR(toPoint: Point): BRAA {
    const deltaX = this.x - toPoint.x
    const deltaY = this.y - toPoint.y

    // distance formula for range
    const rng = Math.floor(
      Math.sqrt(deltaX * deltaX + deltaY * deltaY) / PIXELS_TO_NM
    )

    // convert cartesian angle to heading
    let brg = Math.round(
      270 + toDegrees(Math.atan2(this.y - toPoint.y, this.x - toPoint.x))
    )
    if (brg > 360) {
      brg = brg - 360
    }

    return new BRAA(brg, rng)
  }

  /**
   * Get distance between points (straight along axis, i.e. not slant range)
   */
  public straightDistNM(toPoint: Point, orientation: BlueInThe): number {
    return Math.floor(
      Math.abs(
        orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
          ? this.y - toPoint.y
          : this.x - toPoint.x
      ) / PIXELS_TO_NM
    )
  }
}
