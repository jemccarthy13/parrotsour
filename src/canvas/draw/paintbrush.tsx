import { Braaseye } from "../../classes/braaseye"
import { AircraftGroup } from "../../classes/groups/group"
import { Point } from "../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../utils/psmath"
import { PictureCanvasProps, PictureCanvasState } from "../canvastypes"
import { clampInContext } from "./drawutils"
import { formatAlt } from "./formatutils"

export class PaintBrush {
  private static ctx: CanvasRenderingContext2D

  constructor(ctx: CanvasRenderingContext2D) {
    PaintBrush.ctx = ctx
  }

  public static use(ctx: CanvasRenderingContext2D | undefined | null): void {
    if (ctx) PaintBrush.ctx = ctx
  }

  public static getContext(): CanvasRenderingContext2D {
    return PaintBrush.ctx
  }

  public static clearCanvas(): void {
    const ctx = PaintBrush.ctx
    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height)
  }

  /**
   * Draw a line on the drawing context, given the properties
   * @param startX start x of the line
   * @param startY start y of the line
   * @param endX end x of the line
   * @param endY end y of the line
   * @param color (optional) color of the line, defaults to black
   */
  public static drawLine(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    color = "black",
    forcedCtx?: CanvasRenderingContext2D
  ): void {
    let context = PaintBrush.ctx
    if (forcedCtx) context = forcedCtx
    context.lineWidth = 1
    context.strokeStyle = color
    context.beginPath()
    context.moveTo(startX, startY)
    context.lineTo(endX, endY)
    context.stroke()
    context.stroke()
  }

  /**
   * Draw text to the drawing context at the given position
   * @param text Text to draw
   * @param x X Position to draw at
   * @param y Y Position to draw at
   * @param size (optional) Size of the text to draw, defaults 12
   * @param color (optional) Color of the text, defaults to black
   */
  public static drawText(
    text: string,
    x: number,
    y: number,
    size = 12,
    color = "black"
  ): void {
    const ctx = PaintBrush.ctx
    ctx.lineWidth = 1
    ctx.fillStyle = color
    ctx.font = size + "px Arial"
    const pos = clampInContext(ctx, x, y)

    ctx.fillText(text, pos.x, pos.y)
    ctx.fillText(text, pos.x, pos.y)
  }

  /**
   * Draw a measurement (distance with a length number)
   * @param {Point} start position to start for line
   * @param {Point} end position to end for line
   * @param {number} distance how long the line should be
   * @param {boolean} showMeasurements Display the measurements
   */
  public static drawMeasurement(
    start: Point,
    end: Point,
    distance: number,
    showMeasurements: boolean
  ): void {
    if (showMeasurements) {
      PaintBrush.drawLine(start.x, start.y, end.x, end.y)
      PaintBrush.drawText(
        Math.floor(distance).toString(),
        (start.x + end.x) / 2,
        (start.y + end.y) / 2 - 3
      )
    }
  }

  /**
   * Draw altitudes next to a group with optional offset
   *
   * @param grpPos group's current position
   * @param alts altitudes to draw
   * @param offX (optional) x-axis offset from group
   * @param offY (optional) y-axis offset from group
   */
  public static drawAltitudes(
    grpPos: Point,
    alts: number[],
    offX?: number,
    offY?: number
  ): void {
    const offsetX = offX || 0
    const offsetY = offY || 0
    const formattedAlts: string[] = alts.map((a: number) => {
      return formatAlt(a)
    })
    PaintBrush.drawText(
      formattedAlts.join(","),
      grpPos.x + 25 + offsetX,
      grpPos.y - 11 + offsetY,
      11,
      "#ff8c00"
    )
  }

  public static drawBullseye(bull?: Point, color?: string): Point {
    const context = PaintBrush.ctx
    color = color || "black"
    context.lineWidth = 1
    context.fillStyle = color
    context.strokeStyle = color

    const centerPointX = bull
      ? bull.x
      : randomNumber(context.canvas.width * 0.33, context.canvas.width * 0.66)
    const centerPointY = bull
      ? bull.y
      : randomNumber(context.canvas.height * 0.33, context.canvas.height * 0.66)

    context.beginPath()
    context.arc(
      centerPointX,
      centerPointY,
      PIXELS_TO_NM / 2,
      0,
      2 * Math.PI,
      true
    )
    context.stroke()
    context.fill()

    context.moveTo(centerPointX, centerPointY + PIXELS_TO_NM * 2)
    context.lineTo(centerPointX, centerPointY - PIXELS_TO_NM * 2)
    context.stroke()

    context.moveTo(centerPointX + PIXELS_TO_NM * 2, centerPointY)
    context.lineTo(centerPointX - PIXELS_TO_NM * 2, centerPointY)
    context.stroke()

    return new Point(centerPointX, centerPointY)
  }

  public static drawFullInfo(
    state: PictureCanvasState,
    props: PictureCanvasProps,
    groups: AircraftGroup[]
  ): void {
    for (let y = 0; y < groups.length; y++) {
      const grpPos = groups[y].getCenterOfMass(props.dataStyle)
      if (props.showMeasurements) {
        new Braaseye(
          grpPos,
          state.blueAir.getCenterOfMass(props.dataStyle),
          state.bullseye
        ).draw(true, props.braaFirst)
      }
      PaintBrush.drawAltitudes(grpPos, groups[y].getAltitudes())
    }
  }
}
