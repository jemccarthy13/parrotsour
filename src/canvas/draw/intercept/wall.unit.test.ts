import { SensorType } from "../../../classes/aircraft/datatrail/sensortype"
import { AircraftGroup, GroupParams } from "../../../classes/groups/group"
import { Point } from "../../../classes/point"
import TestCanvas from "../../../testutils/testcanvas"
import { BlueInThe, PictureCanvasState } from "../../canvastypes"
import { testProps } from "./mockutils.unit.test"
import DrawWall from "./wall"
import * as PSMath from "../../../utils/psmath"
import { PaintBrush } from "../paintbrush"

let testState: PictureCanvasState
let p: Partial<GroupParams>
let wall: DrawWall

describe("DrawWall", () => {
  TestCanvas.useContext(800, 500)

  beforeEach(() => {
    PaintBrush.clearCanvas()
  })

  function setBlueInTheNorth() {
    const updatedProps = { ...testProps }
    updatedProps.orientation.orient = BlueInThe.NORTH
    const updatedState = {
      ...testState,
      blueAir: new AircraftGroup({ sx: 200, sy: 50, hdg: 180, nContacts: 4 }),
    }
    wall.initialize(updatedProps, updatedState)
  }

  beforeEach(() => {
    testState = {
      bullseye: new Point(400, 400),
      blueAir: new AircraftGroup({
        sx: 600,
        sy: 400,
        hdg: 270,
        nContacts: 4,
      }),
      answer: { pic: "3 grp ladder", groups: [] },
      reDraw: jest.fn(),
    }

    p = {
      dataTrailType: SensorType.ARROW,
      sx: 200,
      sy: 100,
      nContacts: 4,
      hdg: 90,
      alts: [20, 20, 20, 20],
    }

    testProps.orientation.orient = BlueInThe.EAST
    wall = new DrawWall()
    wall.initialize(testProps, testState)

    jest.restoreAllMocks()
  })

  it("simple_functions", () => {
    const created = new DrawWall().create()
    expect(created.formatPicTitle()).toEqual("0 GROUP WALL")
  })

  it("chooses_num_groups_for_n_contacts", () => {
    wall.chooseNumGroups(1)
    expect(wall.numGroupsToCreate).toEqual(3)
    wall.chooseNumGroups(2)
    expect(wall.numGroupsToCreate).toEqual(3)
    wall.chooseNumGroups(3)
    expect(wall.numGroupsToCreate).toEqual(3)
    jest
      .spyOn(PSMath, "randomNumber")
      .mockReturnValueOnce(4)
      .mockReturnValueOnce(5)
      .mockReturnValueOnce(3)
      .mockReturnValueOnce(5)
    wall.chooseNumGroups(4)
    expect(wall.numGroupsToCreate).toEqual(4)
    wall.chooseNumGroups(5)
    expect(wall.numGroupsToCreate).toEqual(5)
    wall.chooseNumGroups(6)
    expect(wall.numGroupsToCreate).toEqual(3)
    wall.chooseNumGroups(0)
    expect(wall.numGroupsToCreate).toEqual(5)
  })

  it("gets_pic_dimensions", () => {
    jest.spyOn(PSMath, "randomNumber").mockReturnValue(10 * PSMath.PIXELS_TO_NM)
    wall.numGroupsToCreate = 3
    const startPt = new Point(100, 100)
    const pInfo = wall.getPictureInfo(startPt)
    expect(pInfo.wide).toEqual(2 * 10 * PSMath.PIXELS_TO_NM)
    expect(pInfo.deep).toEqual(20 * PSMath.PIXELS_TO_NM)
    expect(pInfo.start).toEqual(startPt)
  })

  it("creates_groups", () => {
    jest.spyOn(PSMath, "randomHeading").mockReturnValue(90)
    jest.spyOn(PSMath, "randomNumber").mockReturnValue(0)
    wall.seps = [0, 20 * PSMath.PIXELS_TO_NM, 20 * PSMath.PIXELS_TO_NM]
    wall.numGroupsToCreate = 3
    const startPt = new Point(100, 100)
    const grps = wall.createGroups(startPt, [1, 2, 3])
    expect(grps[0].getStartPos()).toEqual(startPt)
    expect(grps[0].getHeading()).toEqual(90)
    expect(grps[0].getStrength()).toEqual(1)
    expect(grps[1].getStartPos()).toEqual(new Point(100, 180))
    expect(grps[1].getHeading()).toEqual(90)
    expect(grps[1].getStrength()).toEqual(2)
    expect(grps[2].getStartPos()).toEqual(new Point(100, 260))
    expect(grps[2].getHeading()).toEqual(90)
    expect(grps[2].getStrength()).toEqual(3)
  })

  it("creates_groups_hard_mode", () => {
    const updatedProps = { ...testProps }
    updatedProps.isHardMode = true
    wall.initialize(updatedProps, testState)
    jest
      .spyOn(PSMath, "randomHeading")
      .mockReturnValueOnce(90)
      .mockReturnValueOnce(90)
      .mockReturnValueOnce(120)
      .mockReturnValueOnce(240)
    jest.spyOn(PSMath, "randomNumber").mockReturnValue(0)
    wall.seps = [0, 20 * PSMath.PIXELS_TO_NM, 20 * PSMath.PIXELS_TO_NM]
    wall.numGroupsToCreate = 3
    const startPt = new Point(100, 100)
    const grps = wall.createGroups(startPt, [1, 2, 3])
    expect(grps[0].getStartPos()).toEqual(startPt)
    expect(grps[0].getHeading()).toEqual(90)
    expect(grps[0].getStrength()).toEqual(1)
    expect(grps[1].getStartPos()).toEqual(new Point(100, 180))
    expect(grps[1].getHeading()).toEqual(120)
    expect(grps[1].getStrength()).toEqual(2)
    expect(grps[2].getStartPos()).toEqual(new Point(100, 260))
    expect(grps[2].getHeading()).toEqual(240)
    expect(grps[2].getStrength()).toEqual(3)
  })

  it("creates_groups_ns", () => {
    setBlueInTheNorth()
    jest.spyOn(PSMath, "randomHeading").mockReturnValue(90)
    jest.spyOn(PSMath, "randomNumber").mockReturnValue(0)
    wall.seps = [0, 20 * PSMath.PIXELS_TO_NM, 20 * PSMath.PIXELS_TO_NM]
    wall.numGroupsToCreate = 3
    const startPt = new Point(100, 100)
    const grps = wall.createGroups(startPt, [1, 2, 3])
    expect(grps[0].getStartPos()).toEqual(startPt)
    expect(grps[0].getHeading()).toEqual(90)
    expect(grps[0].getStrength()).toEqual(1)
    expect(grps[1].getStartPos()).toEqual(new Point(180, 100))
    expect(grps[1].getHeading()).toEqual(90)
    expect(grps[1].getStrength()).toEqual(2)
    expect(grps[2].getStartPos()).toEqual(new Point(260, 100))
    expect(grps[2].getHeading()).toEqual(90)
    expect(grps[2].getStrength()).toEqual(3)
  })

  it("draws_info", () => {
    const group1 = new AircraftGroup(p)
    const group2 = new AircraftGroup({ ...p, sy: 180, alts: [13, 13, 13, 13] })
    const group3 = new AircraftGroup({
      ...p,
      sy: 265,
      alts: [15, 15, 15],
      nContacts: 3,
    })
    wall.groups = [group1, group2, group3]
    wall.drawInfo()
    expect(wall.dimensions.wide).toEqual(40)
    expect(TestCanvas.getCanvas()).toMatchSnapshot()
  })

  it("draws", () => {
    TestCanvas.useContext()
    jest.spyOn(PSMath, "randomHeading").mockReturnValue(90)
    jest
      .spyOn(PSMath, "randomNumber")
      .mockReturnValueOnce(3) // num groups
      .mockReturnValueOnce(1) // 1 contact, first group
      .mockReturnValueOnce(2) // 2 contacts, second group
      .mockReturnValueOnce(80) // separation 1-2
      .mockReturnValueOnce(80) // separation 2-3
      .mockReturnValueOnce(0) // heading offset 1
      .mockReturnValueOnce(15) // alt single 1
      .mockReturnValueOnce(50) // mvr chance
      .mockReturnValueOnce(0) // heading offset 2
      .mockReturnValueOnce(17) // alt 2-1
      .mockReturnValueOnce(17) // alt 2-2
      .mockReturnValueOnce(80) // mvr chance
      .mockReturnValueOnce(0) // heading offset 3
      .mockReturnValueOnce(20) // alt 3-1
      .mockReturnValueOnce(20) // alt 3-2
      .mockReturnValueOnce(20) // alt 3-3
      .mockReturnValueOnce(20) // alt 3-4
      .mockReturnValueOnce(0) // mvr chance
    wall.draw(false, 7, new Point(100, 100))
    expect(TestCanvas.getCanvas()).toMatchSnapshot()

    expect(wall.groups[0].doesManeuvers()).toEqual(true)
    expect(wall.groups[1].doesManeuvers()).toEqual(false)
    expect(wall.groups[2].doesManeuvers()).toEqual(false)

    expect(wall.formatWeighted()).toEqual("")
  })
})
