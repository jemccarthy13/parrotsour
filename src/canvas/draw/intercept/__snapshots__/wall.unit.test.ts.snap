// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawWall draws 1`] = `
<canvas
  data-snapshot-image="a0034dfe27a569e05085f4eedc47666fcbe65fa3932158a1487381562abf5ec138c93c7e95ba6f96a8d074f2cc43b18950c19e51a41ba238e956bd8dc0f425bc"
  height="500"
  width="800"
/>
`;

exports[`DrawWall draws_info 1`] = `
<canvas
  data-snapshot-image="12ad573ed4af88faf33751a75c0376b0d4aca60dff07051c7393f5362216d249fb72726cce64b7503a44ed6ec21f4790c0f6973818497845ecb96eea592da6fc"
  height="500"
  width="800"
/>
`;
